src/emitter/batch_emitter.rs

```diff
@@ -163,43 +180,75 @@ impl BatchEmitter {
                 // or the channel is closed and there are no more messages, in which case we exit the loop
-                let message = match rx.recv().await {
-                    Some(msg) => {
-                        log::debug!("Received message: {:?}", msg);
-                        msg
-                    }
+
+                // select! is used to check both the `retry_rx` channel and the `rx` channel for new messages
+                let message = match tokio::select! {
+                    // `biased;` is used to ensure that the `retry_rx` channel is checked first, so retries get priority
+                    biased;
+
+                    retry = retry_rx.recv() => retry,
+                    event = rx.recv() => event,
+                } {
+                    Some(message) => message,
                     None => break,
                 };
 
                 match message {
-                    EmitterMessage::Send(batch) => {
-                        // Clone HttpClient to be moved into the task
+                    EmitterMessage::Send(mut batch) => {
+                        // Clone HttpClient and the retry transmitter to be moved into the task
                         let client = http_client.clone();
+                        let retry_transmitter = retry_tx.clone();
 
                         // Spawn a new task to send the batch
                         tokio_tasks.push(tokio::spawn(async move {
+                            if let Some(delay) = batch.delay {
+                                log::debug!("Delaying batch {} for {:?}", batch.id, delay);
+                                tokio::time::sleep(delay).await;
+
+                                if let Err(e) = batch.update_event_stm() {
+                                    // If the update fails, we just re-send the batch as-is
+                                    // Not ideal, but it's better than losing events
+                                    log::warn!(
+                                        "Failed to update stm of events in batch {} for retry: {e}",
+                                        batch.id
+                                    )
+                                };
+                            };
+
                             let batch_length = batch.events.len();
                             match Self::send_batch(batch, client).await {
-                                Ok(_) => {
-                                    log::info!("Sent batch of {batch_length} events")
+                                Ok(batch_id) => {
+                                    log::info!("Sent batch {batch_id} of {batch_length} events")
+                                }
+                                Err(mut failed_batch) => {
+                                    failed_batch.update_delay();
+
+                                    let batch_id = failed_batch.id;
+                                    match retry_transmitter.send(EmitterMessage::Send(failed_batch))
```

Comment:
I am a bit worried that there is no cap on the number of retries here. In a long running and busy server, it may happen that when the collector is down for a longer period of time, a large number of retry attempts will be scheduled which may cause the server app to run out of memory and crash (or am I missing something about the implementation here?). I can see two ways to address this:

1. Add a limit on the max number of retries for each batch. So the `EventBatch` object could have an internal counter which would be checked before deciding whether to retry again. This is [implemented in the Scala tracker, see here](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/scala-tracker/initialization/#retry-policies).
2. (this is not relevant for the implementation here, just adding for completeness) Add a limit on the max size of the event queue. In case events to be retried are added back to the event queue, there may be a limit on the queue size which would drop events from the queue in case it is reached (in FIFO or LIFO fashion). This is implemented in the Java tracker for instance.